Translated the index warnings / errors.

joachimnielandt opened the pull request:
Noticed some warnings / errors that were not translated yet. This change was implemented in the VL fork.

I modified the indexed for this purpose, storing an object instead of a keyword. This makes querying specifically for warnings easier, but allows for translating the strings by using the embedded values.c

Not all schema plugins were modified - I will update them if the approach is deemed OK.

Example:
![image](https://github.com/geonetwork/core-geonetwork/assets/11455912/c8294dda-7d0f-4dd7-ac92-518584ac4dfb)


cc. @fxprunayre 
  
<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geonetwork/core-geonetwork/blob/main/CONTRIBUTING.md
- [x] *Pull request* provided for `main` branch, backports managed with label
- [ ] *Good housekeeping* of code, cleaning up comments, tests, and documentation
- [ ] *Clean commit history* broken into understandable chucks, avoiding big commits with hundreds of files, cautious of reformatting and whitespace changes
- [ ] *Clean commit message*s, longer verbose messages are encouraged
- [ ] *API Changes* are identified in commit messages
- [ ] *Testing* provided for features or enhancements using [automatic tests](https://github.com/geonetwork/core-geonetwork/blob/main/software_development/TESTING.md))
- [ ] *User documentation* provided for new features or enhancements in [mannual](https://github.com/geonetwork/core-geonetwork/tree/main/docs/manual)
- [ ] *Build documentation* provided for development instructions in `README.md` files
- [ ] *Library management* using `pom.xml` dependency management. Update build documentation with intended library use and library tutorials or documentation

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

fxprunayre commented:
The approach sounds good to me @joachimnielandt

fxprunayre commented on schemas/iso19115-3.2018/src/main/plugin/iso19115-3.2018/index-fields/index.xsl:
```diff
@@ -338,7 +338,16 @@
+                  "string": "indexingErrorMsg-invalidDateFormat",
+                  "type": "warning",
+                  "values": {
+                  "dateType": "<xsl:value-of select="$dateType"/>",
+                  "date": "<xsl:value-of select="$date"/>"
```
Use `gn-fn-index:json-escape` to be sure to create valid JSON
```suggestion
                  "dateType": "<xsl:value-of select="gn-fn-index:json-escape($dateType)"/>",
                  "date": "<xsl:value-of select="gn-fn-index:json-escape($date)"/>"
```

fxprunayre commented on web/src/main/webResources/WEB-INF/data/config/index/records.json:
```diff
@@ -1990,7 +1990,18 @@
+        "type": "object",
+        "properties": {
+          "string": {
+            "type": "text"
```
Can be `keyword` because it is a key and we don't run full text search on it.

```suggestion
            "type": "keyword"
```

it will also be used as a facet on admin console > status

joachimnielandt commented:
> The approach sounds good to me @joachimnielandt

The only remaining reference I see to the <indexingErrorMsg> tag is the dcat2 plugin, but that's one of our local additions. I think this PR then covers all schema plugins?

fxprunayre commented:
Some work required in the admin @joachimnielandt around https://github.com/geonetwork/core-geonetwork/blob/main/web-ui/src/main/resources/catalog/js/admin/DashboardStatusController.js#L327. Can you check http://localhost:8080/geonetwork/srv/eng/admin.console?debug#/dashboard

Also one fix for thesaurus field https://github.com/Informatievlaanderen/metadata-geonetwork/pull/1

And can you rebase on main to align with what has been done in https://github.com/geonetwork/core-geonetwork/pull/7525

joachimnielandt commented:
> Some work required in the admin @joachimnielandt around https://github.com/geonetwork/core-geonetwork/blob/main/web-ui/src/main/resources/catalog/js/admin/DashboardStatusController.js#L327. Can you check http://localhost:8080/geonetwork/srv/eng/admin.console?debug#/dashboard
> 
> Also one fix for thesaurus field [Informatievlaanderen#1](https://github.com/Informatievlaanderen/metadata-geonetwork/pull/1)
> 
> And can you rebase on main to align with what has been done in #7525

Rebased on latest core/main, attempted to tie up the last references to the indexingErrorMsg object that you mentioned.

Before, I think the message was hardcoded to contain pipe characters (didn't see those in my previous tests though), based on that some styling was changed. I simplified this to:
- show the indexingErrorMsg type (warning / error / ...?)
- translate the indexingErrorMsg string, complete with values sent
![image](https://github.com/geonetwork/core-geonetwork/assets/11455912/dd0c3d54-cc06-4121-81e8-f625b77900fd)

joachimnielandt commented on web-ui/src/main/resources/catalog/components/elasticsearch/EsFacet.js:
```diff
@@ -232,16 +232,16 @@
+                field: "indexingErrorMsg.type",
                 size: 10,
-                exclude: "Warning.*"
+                exclude: "warning"
```
I spotted these too @fxprunayre , have to be modified too I reckon?

fxprunayre replied:
I would replace `indexingErrorMsg` and `indexingWarningMsg` by 
```
indexingErrorMsg: {
              terms: {
                field: "indexingErrorMsg.string",
                size: 10
              }
            }
```
in order to display each types of error or warning to easily filter on it

![image](https://github.com/geonetwork/core-geonetwork/assets/1701393/d58021d9-5aaa-432f-b350-d1504ff2914f)

What do you think?

joachimnielandt replied:
That makes sense to me. I have pushed your proposed change, but could not verify that these facets are actually shown. 

How to configure the search / editor board to show this facet? (I do not see "error types" and "indexing errors" as shown in your screenshot).

fxprunayre replied:
This facet is displayed in admin > status - in the indexing error block

joachimnielandt replied:
Excellent, thanks!

joachimnielandt replied:
Verified this is working. The only drawback here is that the values are not replaced in the `indexingErrorMsg.string` template should they be present. But doing that would also create a lot of duplicate facet values in the list. Perhaps best to keep it to the template string then.

joachimnielandt commented:
@fxprunayre I have pushed an extra change that follows your json input sanitation, to make sure only valid values end up in the `indexingErrorMsg` object. Additionally, I have rebased on the latest `main` version.

joachimnielandt commented:
Due to switching forks and dealing with missing schemas I found some other places where indexingErrorMsg is pushed to the index. I have added another commit to deal with this. Some examples of my tests that trigger these failures, with the resulting index field shown:

**example of index field for style sheet error:**
```json
"indexingErrorMsg": [
    {
        "string": "indexingErrorMsg-indexingStyleSheetError",
        "values": {
            "message": "Failed to compile stylesheet. 1 error detected."
        },
        "type": "error"
    }
],
```

**example of unregistered schema:**
```json
"indexingErrorMsg": [
    {
        "string": "indexingErrorMsg-schemaNotRegistered",
        "values": {
            "schema": "dcat2",
            "record": "112"
        },
        "type": "error"
    }
],
```

fxprunayre reviewed (approved):
LGTM @joachimnielandt. I fixed conflict related to Elastic 8 update in https://github.com/Informatievlaanderen/metadata-geonetwork/pull/3
